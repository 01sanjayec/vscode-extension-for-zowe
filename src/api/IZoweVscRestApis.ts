/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import * as zowe from "@brightside/core";
import * as imperative from "@brightside/imperative";

/**
 * This namespace provides interfaces for all the external APIs provided by this VS Code Extension.
 * Other VS Code Extension can implement these and use the IApiRegister interface to register themselves.
 */
export namespace ZoweVscApi {
    /**
     * Common interface shared between all API interfaces offered by this extension.
     */
    export interface ICommon {
        /**
         * Return the type name of the CLI profile supported by this api.
         */
        getProfileTypeName(): string;

        /**
         * Create a session for the specific profile type
         * @param profile {imperative.IProfile} profile reference
         * @returns {imperative.Session} a Zowe CLI Session reference
         */
        createSession(profile: imperative.IProfile): imperative.Session;
    }

    /**
     * API for providing a USS Rest handler to the extension.
     * @export
     * @interface IUss
     */
    export interface IUss extends ICommon {
        /**
         * Return the directory elements for a given USS path.
         *
         * @param {string} path
         * @returns {IZosFilesResponse}
         *     A response structure that contains a boolean success property
         *     as well as the list of results in apiResponse.items with
         *     minimal properties name, mode.
         */
        fileList(
            session: imperative.Session,
            path: string
        ): Promise<zowe.IZosFilesResponse>;

        /**
         * Check th USS chtag.
         *
         * @param {imperative.Session} session
         * @param {string} USSFileName
         * @returns {Promise<boolean>}
         */
        isFileTagBinOrAscii(
            session: imperative.Session,
            USSFileName: string
        ): Promise<boolean>;

        /**
         * Retrieve the contents of a USS file.
         * @param {imperative.Session} session
         * @param {string} ussFileName
         * @param {zowe.IDownloadOptions} options
         */
        getContents(
            session: imperative.Session,
            ussFileName: string,
            options: zowe.IDownloadOptions
        ): Promise<zowe.IZosFilesResponse> ;

        /**
         * Uploads the files at the given path. Use for Save.
         *
         * @param {imperative.Session} session
         * @param {string} inputFile
         * @param {string} ussname
         * @param {boolean} [binary]
         * @param {string} [localEncoding]
         * @returns {Promise<zowe.IZosFilesResponse>}
         */
        putContents(
            session: imperative.Session,
            inputFile: string,
            ussname: string,
            binary?: boolean,
            localEncoding?: string
        ): Promise<zowe.IZosFilesResponse>;

        /**
         * Create a new directory or file in the specified path.
         *
         * @param {imperative.Session} session
         * @param {string} ussPath
         * @param {string} type
         * @param {string} [mode]
         * @returns {Promise<string>}
         */
        create(
            session: imperative.Session,
            ussPath: string,
            type: string,
            mode?: string
        ): Promise<string>;

        /**
         * Deletes the USS file at the given path.
         *
         * @param {imperative.Session} session
         * @param {string} fileName
         * @param {boolean} [recursive]
         * @returns {Promise<zowe.IZosFilesResponse>}
         * @memberof IUss
         */
        delete(
            session: imperative.Session,
            fileName: string,
            recursive?: boolean): Promise<zowe.IZosFilesResponse>;
    }

    // TODO
    export interface IMvs extends ICommon {
        dataSet(
            session: imperative.Session,
            filter: string
        ): Promise<zowe.IZosFilesResponse>;
        allMembers(
            session: imperative.Session,
            dataSetName: string
        ): Promise<zowe.IZosFilesResponse>;
    }

    export interface IApiRegister {

        /**
         * Register a new implementation of the USS Api.
         *
         * @param {IUss} ussApi
         * @memberof IZoweVscApiRegister
         */
        registerUssApi(ussApi: IUss): void;

        /**
         * Lookup of an API for USS for a given profile.
         * @private
         * @param {string} profileType
         * @returns
         * @memberof ZoweVscApiRegister
         */
        getUssApi(profile: imperative.IProfileLoaded): IUss;
    }
}
