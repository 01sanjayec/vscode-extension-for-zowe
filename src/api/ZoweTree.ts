/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import * as vscode from "vscode";
import { Session } from "@brightside/imperative";
import { IJob } from "@brightside/core";

/**
 * The base interface for Zowe tree browsers that implement the
 * vscode.TreeDataProvider.
 *
 * @export
 * @interface IZoweTree
 * @extends {vscode.TreeDataProvider<T>}
 * @template T provide a subtype of vscode.TreeItem
 */
export interface IZoweTree<T> extends vscode.TreeDataProvider<T> {
    /**
     * Root session nodes
     */
    mSessionNodes: T[];
    /**
     * Root favorites node
     */
    mFavoriteSession: T;
    /**
     * Array of favorite nodes
     */
    mFavorites: T[];
    /**
     * Adds a session to the container
     * @param sessionName
     * @param type e.g. zosmf
     */
    addSession(sessionName?: string, type?: string): Promise<void>;
    /**
     * Adds a favorite node
     * @param favorite Adds a favorite node
     */
    addFavorite(favorite: IZoweTreeNode);
    /**
     * Removes a favorite node
     * @param favorite Adds a favorite node
     */
    removeFavorite(node: IZoweTreeNode);
    /**
     * Refreshes the tree
     */
    refresh(): void;
    /**
     * Refreshes an element of the tree
     * @param favorite Node to refresh
     */
    refreshElement(node: IZoweTreeNode): void;
}
/**
 * The base interface for Zowe tree nodes.
 *
 * @export
 * @interface IZoweTreeNode
 */
export interface IZoweTreeNode {
    /**
     * The icon path or [ThemeIcon](#ThemeIcon) for the tree item.
     */
    iconPath?: string | vscode.Uri | { light: string | vscode.Uri; dark: string | vscode.Uri } | vscode.ThemeIcon;
    /**
     * Indicator that the child data may have become stale and requires refreshing.
     */
    dirty: boolean;
    /**
     *  A human-readable string describing this item.
     */
    label?: string;
    /**
     * The tooltip text when you hover over this item.
     */
    tooltip?: string;
    /**
     * Describes the full path of a file
     */
    fullPath?: string;
    /**
     * Children nodes of this node
     */
    children?: IZoweTreeNode[];
    /**
     * [TreeItemCollapsibleState](#TreeItemCollapsibleState) of the tree item.
     */
    collapsibleState?: vscode.TreeItemCollapsibleState;
    /**
     * Context value of the tree item. This can be used to contribute item specific actions in the tree.
     *
     * This will show action `extension.deleteFolder` only for items with `contextValue` is `folder`.
     */
    contextValue?: string;
    /**
     * Retrieves the node label
     */
    getLabel(): string;
    /**
     * Retrieves the nodes parent node
     */
    getParent(): IZoweTreeNode;
    /**
     * Retrieves the nodes children nodes
     */
    getChildren(): Promise<IZoweTreeNode[]>;
    /**
     * Retrieves the profile name in use with this node
     */
    getProfileName(): string;
    /**
     * Retrieves the session node in use with this node
     */
    getSessionNode(): IZoweTreeNode;
    /**
     * Retrieves the session object in use with this node
     */
    getSession(): Session;
}

/**
 * Extended interface for Zowe Dataset tree nodes.
 *
 * @export
 * @interface export interface IZoweDatasetTreeNode extends IZoweTreeNode {
 */
export interface IZoweDatasetTreeNode extends IZoweTreeNode {
    /**
     * Search criteria for a Dataset search
     */
    pattern?: string;

    /**
     * Retrieves child nodes of this IZoweDatasetTreeNode
     *
     * @returns {Promise<IZoweDatasetTreeNode[]>}
     */
    getChildren(): Promise<IZoweDatasetTreeNode[]>;
    /**
     * Retrieves the etag value for the file
     *
     * @returns {string}
     */
    getEtag?(): string;
    /**
     * Sets the etag value for the file
     *
     * @param {string}
     */
    setEtag?(etag: string);
}

/**
 * Extended interface for Zowe USS tree nodes.
 *
 * @export
 * @interface export interface IZoweUSSTreeNode extends IZoweTreeNode {
 */
export interface IZoweUSSTreeNode extends IZoweTreeNode {
    /**
     * Retrieves an abridged for of the label
     */
    shortLabel?: string;
    /**
     * List of child nodes downloaded in binary format
     */
    binaryFiles?: {};
    /**
     * Binary indicator. Default false (text)
     */
    binary?: boolean;
    /**
     * Specific profile name in use with this node
     */
    mProfileName?: string;

    /**
     * Retrieves child nodes of this IZoweUSSTreeNode
     *
     * @returns {Promise<IZoweUSSTreeNode[]>}
     */
    getChildren(): Promise<IZoweUSSTreeNode[]>;
    /**
     * Retrieves the etag value for the file
     *
     * @returns {string}
     */
    getEtag?(): string;
    /**
     * Sets the etag value for the file
     *
     * @param {string}
     */
    setEtag?(etag: string);
    /**
     * Renaming a USS Node. This could be a Favorite Node
     *
     * @param {string} newNamePath
     */
    rename?(newNamePath: string);
}

/**
 * Extended interface for Zowe Job tree nodes.
 *
 * @export
 * @interface export interface IZoweJobTreeNode extends IZoweTreeNode {
 */
export interface IZoweJobTreeNode extends IZoweTreeNode {

    /**
     * Standard job response document
     * Represents the attributes and status of a z/OS batch job
     * @interface IJob
     */
    job?: IJob;
    /**
     * Search criteria for a Job search
     */
    searchId?: string;
    /**
     * Job Prefix i.e "MYJOB"
     * Attribute of Job query
     */
    prefix?: string;
    /**
     * Job Owner i.e "MYID"
     * Attribute of Job query
     */
    owner?: string;

    /**
     * Retrieves child nodes of this IZoweJobTreeNode
     *
     * @returns {Promise<IZoweJobTreeNode[]>}
     */
    getChildren(): Promise<IZoweJobTreeNode[]>;
}
