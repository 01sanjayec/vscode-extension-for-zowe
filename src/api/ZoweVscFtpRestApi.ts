/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import * as ftp from "@brightside/core";  // import the ftp plugin REST API instead
import * as imperative from "@brightside/imperative";

import { IZoweVscUssApi } from "./IZoweVscRestApis";

export class ZoweVscFtpUssRestApi implements IZoweVscUssApi {

    public getProfileTypeName(): string {
        return "zftp";
    }

    public createSession(profile: imperative.IProfile): imperative.Session {
        // ftp.FtpSession.createBasicFtpSession(profile);
        // do the mapping and return
        return undefined;
    }

    public async fileList(session: imperative.Session, path: string): Promise<ftp.IZosFilesResponse>{
        return ftp.List.fileList(session, path);
    }

    public async create(session: imperative.Session, ussPath: string, type: string, mode?: string): Promise<string> {
        return undefined;
    }
}
