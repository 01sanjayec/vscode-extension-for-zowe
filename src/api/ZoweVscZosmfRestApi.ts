/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import * as zowe from "@brightside/core";
import * as imperative from "@brightside/imperative";

import { IZoweVscUssApi } from "./IZoweVscRestApis";

export class ZoweVscZosmfUssRestApi implements IZoweVscUssApi {

    public getProfileTypeName(): string {
        return "zosmf";
    }
    public createSession(profile: imperative.IProfile): imperative.Session {
        return zowe.ZosmfSession.createBasicZosmfSession(profile);
    }

    public async fileList(session: imperative.Session, path: string): Promise<zowe.IZosFilesResponse> {
        return zowe.List.fileList(session, path);
    }

    public async create(session: imperative.Session, ussPath: string, type: string, mode?: string): Promise<string> {
        return zowe.Create.uss(session, ussPath, type);
    }
}
