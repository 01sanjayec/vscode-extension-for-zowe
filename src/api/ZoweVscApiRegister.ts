/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import * as zowe from "@brightside/core";
import * as imperative from "@brightside/imperative";

import { IZoweVscUssApi, IZoweVscMvsApi } from "./IZoweVscRestApis";
import { ZoweVscZosmfUssRestApi } from "./ZoweVscZosmfRestApi";

/**
 * The API register that gets exposed to other VS Code extensions to contribute their implementations.
 * @export
 * @class ZoweVscApiRegister
 */
export class ZoweVscApiRegister {

    /**
     * Access the singleton instance.
     *
     * @static
     * @returns
     * @memberof ZoweVscApiRegister
     */
    public static getInstance() {
        return ZoweVscApiRegister.register;
    }
    /**
     * This object represents the API that gets exposed to other VS Code extensions
     * that want to contribute alternative implementations such as alternative ways
     * of retrieving files and data from z/OS.
     */
    private static register: ZoweVscApiRegister = new ZoweVscApiRegister();

    // These are the different API registries available to extenders
    private zoweVscUssApiImplementations = new Map<string, IZoweVscUssApi>();
    private zoweVscMvsApiImplementations = new Map<string, IZoweVscMvsApi>(); // TODO

    /**
     * Creates an instance of ZoweVscApiRegister.
     * @param {IZoweVscUssApi} [ussApi] Optional. If none provided then it will be initialized with a zOSMF Api.
     */
    private constructor() {
        this.registerUssApi(new ZoweVscZosmfUssRestApi());
    }

    /**
     * Other VS Code extension need to call this to register their USS APIs.
     * @param {IZoweVscUssApi} ussApi
     */
    public registerUssApi(ussApi: IZoweVscUssApi): void {
        if (ussApi && ussApi.getProfileTypeName()) {
            this.zoweVscUssApiImplementations.set(ussApi.getProfileTypeName(), ussApi);
        } else {
            throw new Error("A Zowe Extension client tried to register an invalid API.");
        }
    }

    /**
     * Get an array of all the registered APIs identified by the CLI profile types,
     * such as ["zosmf", "ftp"].
     * @returns {string[]}
     */
    public registeredApiTypes(): string[] {
        return [...new Set([...this.registeredUssApiTypes(), ...this.registeredMvsApiTypes()])];
    }

    /**
     * Get an array of all the registered USS APIs identified by the CLI profile types,
     * such as ["zosmf", "ftp"].
     * @returns {string[]}
     */
    public registeredUssApiTypes(): string[] {
        return [...this.zoweVscUssApiImplementations.keys()];
    }

    /**
     * Get an array of all the registered MVS APIs identified by the CLI profile types,
     * such as ["zosmf", "zftp"].
     * @returns {string[]}
     */
    public registeredMvsApiTypes(): string[] {
        return [...this.zoweVscMvsApiImplementations.keys()];
    }

    // ** Start Common API Methods
    public createSession(profile: imperative.IProfileLoaded): imperative.Session {
        return this.getUssApi(profile).createSession(profile.profile);
    }

    // ** Start of USS Methods

    public async fileList(
        profile: imperative.IProfileLoaded,
        session: imperative.Session,
        path: string): Promise<zowe.IZosFilesResponse>{
        return this.getUssApi(profile).fileList(session, path);
    }

    public async create(
        profile: imperative.IProfileLoaded,
        session: imperative.Session,
        ussPath: string,
        type: string,
        mode?: string): Promise<string> {
        return this.getUssApi(profile).create(session, ussPath, type, mode);
    }

    // ** Private utility methods

    /**
     * Shortcut to look up an API for USS.
     * @private
     * @param {string} profileType
     * @returns
     * @memberof ZoweVscApiRegister
     */
    private getUssApi(profile: imperative.IProfileLoaded): IZoweVscUssApi {
        if (profile && profile.type && this.registeredUssApiTypes().includes(profile.type)) {
            return this.zoweVscUssApiImplementations.get(profile.type);
        }
        else {
            throw new Error("Internal error: Tried to call a non-existing API: " + profile.type);
        }
    }
}
