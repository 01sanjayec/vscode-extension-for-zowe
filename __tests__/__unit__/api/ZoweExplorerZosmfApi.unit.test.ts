/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { ZosmfUssApi, ZosmfMvsApi } from "../../../src/api/ZoweExplorerZosmfApi";
import * as zowe from "@zowe/cli";

describe("Zosmf API tests", () => {

    it("should test that copy data set uses default options", async () => {

        const dataSet = jest.fn(async (session, toDataSet, options) => {
            expect(options).toMatchSnapshot();
            return { api: "", commandResponse: "", success: true };
        });

        (zowe as any).Copy = { dataSet };

        const api = new ZosmfMvsApi();
        api.getSession = jest.fn();
        await api.copyDataSetMember({ dataSetName: "IBM.FROM", memberName: "IEFBR14" }, { dataSetName: "IBM.TO", memberName: "IEFBR15" });

    });

    it("should test that copy data set uses enq", async () => {

        const dataSet = jest.fn(async (session, toDataSet, options) => {
            expect(options).toMatchSnapshot();
            return { api: "", commandResponse: "", success: true };
        });

        (zowe as any).Copy = { dataSet };

        const api = new ZosmfMvsApi();
        api.getSession = jest.fn();
        await api.copyDataSetMember({ dataSetName: "IBM.FROM", memberName: "IEFBR14" }, { dataSetName: "IBM.TO", memberName: "IEFBR15" },
            { enq: "SHR", fromDataSet: { dataSetName: "BROADCOM.FROM" } });

    });

    it("should test that copy data set uses enq only", async () => {

        const dataSet = jest.fn(async (session, toDataSet, options) => {
            expect(options).toMatchSnapshot();
            return { api: "", commandResponse: "", success: true };
        });

        (zowe as any).Copy = { dataSet };

        const api = new ZosmfMvsApi();
        api.getSession = jest.fn();
        await api.copyDataSetMember({ dataSetName: "IBM.FROM", memberName: "IEFBR14" }, { dataSetName: "IBM.TO", memberName: "IEFBR15" },
            { enq: "SHR" } as any);

    });

    it("should test that common putContent is called by putContents", async () => {

        const api = new ZosmfUssApi();

        (api.putContent as any) = jest.fn<ReturnType<typeof api.putContents>, Parameters<typeof api.putContents>>(
            async (inputFilePath: string, ussFilePath: string,
                binary?: boolean, localEncoding?: string,
                etag?: string, returnEtag?: boolean) => {

                return {
                    success: true,
                    commandResponse: "whatever"
                };

            });

        await api.putContents("someLocalFile.txt", "/some/remote", true);

        expect(api.putContent).toBeCalledTimes(1);
        expect(api.putContent).toBeCalledWith("someLocalFile.txt", "/some/remote", {
            binary: true,
        });

    });

});
