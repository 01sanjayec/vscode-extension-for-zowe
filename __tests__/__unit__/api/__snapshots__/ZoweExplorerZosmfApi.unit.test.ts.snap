// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zosmf API tests should test that copy data set uses default options 1`] = `
Object {
  "fromDataSet": Object {
    "dataSetName": "IBM.FROM",
    "memberName": "IEFBR14",
  },
}
`;

exports[`Zosmf API tests should test that copy data set uses enq 1`] = `
Object {
  "enq": "SHR",
  "fromDataSet": Object {
    "dataSetName": "BROADCOM.FROM",
  },
}
`;

exports[`Zosmf API tests should test that copy data set uses enq only 1`] = `
Object {
  "enq": "SHR",
  "fromDataSet": Object {
    "dataSetName": "IBM.FROM",
    "memberName": "IEFBR14",
  },
}
`;
